cpp/src/reductions/reduction_dispatcher_multistep.cuh

```diff
@@ -0,0 +1,146 @@
+ */
+
+#ifndef CUDF_REDUCTION_DISPATCHER_MULTISTEP_CUH
+#define CUDF_REDUCTION_DISPATCHER_MULTISTEP_CUH
+
+#include "reduction_functions.cuh"
+
+namespace cudf {
+namespace reduction {
+namespace detail {
+
+/** --------------------------------------------------------------------------*
+ * @brief Reduction for mean, var, std
+ * It requires extra step after single step reduction call
+ *
+ * @param[in] col    input column
+ * @param[out] scalar  output scalar data
+ * @param[in] ddof   `Delta Degrees of Freedom` used for `std`, `var`.
+ *                   The divisor used in calculations is N - ddof, where N represents the number of elements.
+ * @param[in] stream cuda stream
+ *
+ * @tparam ElementType  the input column cudf dtype
+ * @tparam ResultType   the output cudf dtype
+ * @tparam Op           the operator of cudf::reduction::op::
+ * @tparam has_nulls    true if column has nulls
+ * ----------------------------------------------------------------------------**/
+ template<typename ElementType, typename ResultType, typename Op, bool has_nulls>
+void compound_reduction(gdf_column const& col, gdf_scalar& scalar,
+    gdf_size_type ddof, cudaStream_t stream)
+{
+    gdf_size_type valid_count = col.size - col.null_count;
+
+    ResultType identity = Op::Op::template identity<ResultType>();
+    using intermediateOp = typename Op::template intermediate<ResultType>;
+    // IntermediateType: intermediate structure, output type of `reduction_op` and
+    // input type of `intermediateOp::ComputeResult`
+    using IntermediateType = typename intermediateOp::IntermediateType;
+    IntermediateType intermediate{0};
+
+    // allocate temporary memory for the dev_result
+    void *dev_result = NULL;
+    RMM_TRY(RMM_ALLOC(&dev_result, sizeof(IntermediateType), stream));
+
+    // initialize output by identity value
+    CUDA_TRY(cudaMemcpyAsync(dev_result, &intermediate,
+            sizeof(IntermediateType), cudaMemcpyHostToDevice, stream));
+    CHECK_STREAM(stream);
+
+    // reduction by iterator
+    auto it = Op::template make_iterator<has_nulls, ElementType, ResultType>(col, identity);
+    reduce(static_cast<IntermediateType*>(dev_result), it, col.size, intermediate,
+        typename Op::Op{}, stream);
+
+    // read back the dev_result to host memory
+    // TODO: asynchronous copy
+    CUDA_TRY(cudaMemcpy(&intermediate, dev_result,
+            sizeof(IntermediateType), cudaMemcpyDeviceToHost));
+
+    // compute the dev_result value from intermediate value.
+    ResultType hos_result = intermediateOp::compute_result(intermediate, valid_count, ddof);
+    memcpy(&scalar.data, &hos_result, sizeof(ResultType));
+
+    // cleanup temporary memory
+    RMM_TRY(RMM_FREE(dev_result, stream));
+
+    // set scalar is valid
+    scalar.is_valid = true;
+};
+
+// @brief result type dispatcher for compound reduction (a.k.a. mean, var, std)
+template <typename ElementType, typename Op>
+struct compound_reduction_result_type_dispatcher {
```

Comment:
`compound_reduction_result_type_dispatcher` and `compound_reduction_element_type_dispatcher` are redundant with their `simple_reduction` counterparts. The only difference is in invoking `compound_reduction` vs `simple_reduction`.

These type dispatched functors should be consolidated into simply `dispatch_element_type` and `dispatch_result_type` functors where you control the invoked reduction function by passing in a functor for the reduction function.

Reply:
> `compound_reduction_result_type_dispatcher` and `compound_reduction_element_type_dispatche`r are redundant with their simple_reduction counterparts. The only difference is in invoking `compound_reduction` vs `simple_reduction`.

No, the implementation of is_supported_v() are different between `compound` and `simple`

https://github.com/rapidsai/cudf/blob/13901692b8aecb7fbcd94fe177134ea1de67be3f/cpp/src/reductions/reduction_dispatcher.cuh#L75-L85
https://github.com/rapidsai/cudf/blob/13901692b8aecb7fbcd94fe177134ea1de67be3f/cpp/src/reductions/reduction_dispatcher_multistep.cuh#L89-L94


However, I agree the name are redundant. I feel we should rename the name space from `cudf::reduction::detail` => `cudf::reduction::compound` and  `cudf::reduction::simple`.
How do you think?

Reply:
Ah, you're right, I definitely missed the difference in `is_supported_v`. 

The new naming sounds good to me.

Reply:
Done